### PR TITLE
Heavily increased performance

### PR DESCRIPTION
**Pull request by guassmith**
Modified instanced rendering to map a buffer instead of updating a uniform array every frame

### CODE REVIEW

File: src/main/java/graphics/opengl/OpenGLUtils.java [hunk 1]
```diff
@@ -256,4 +274,25 @@ public static void drawTextRed(String text, float x, float y) {
...
+		if(glError != 0) {
+			if(glError == GL_INVALID_ENUM) errorName = "GL_INVALID_ENUM";
+			if(glError == GL_INVALID_VALUE) errorName = "GL_INVALID_VALUE";
+			if(glError == GL_INVALID_OPERATION) errorName = "GL_INVALID_OPERATION";
+			if(glError == GL_STACK_OVERFLOW) errorName = "GL_STACK_OVERFLOW";
+			if(glError == GL_STACK_UNDERFLOW) errorName = "GL_STACK_UNDERFLOW";
+			if(glError == GL_OUT_OF_MEMORY) errorName = "GL_OUT_OF_MEMORY";
+			if(glError == GL_INVALID_FRAMEBUFFER_OPERATION) errorName = "GL_INVALID_FRAMEBUFFER_OPERATION";
+			if(glError == GL_CONTEXT_LOST) errorName = "GL_CONTEXT_LOST";
+			if(glError == GL_TABLE_TOO_LARGE) errorName = "GL_TABLE_TOO_LARGE";
```
Comment (Bramelvix) [inline src/main/java/graphics/opengl/OpenGLUtils.java:289]: Maybe make a switch here

Comment (guassmith) [reply src/main/java/graphics/opengl/OpenGLUtils.java:289]: sure but performance doesn't matter here, make it so it looks nice

File: src/main/java/map/Level.java [hunk 2]
```diff
@@ -309,23 +309,17 @@ public void render(Vec2f scroll) {
...
 
-		//System.out.println(tileRenderList.size());
-		OpenGLUtils.drawTiles(tileRenderList, new Vec2f(Sprite.SIZE), rowLength, scroll, SpritesheetHashtable.get(1).getId());
-		OpenGLUtils.drawTiles(entityRenderList, new Vec2f(Sprite.SIZE), rowLength, scroll, SpritesheetHashtable.get(1).getId());
-		OpenGLUtils.drawTiles(itemRenderList, new Vec2f(Sprite.SIZE), rowLength, scroll, SpritesheetHashtable.get(1).getId());
+		OpenGLUtils.drawInstanced(OpenGLUtils.tileData, new Vec2f(Sprite.SIZE), scroll);
+		if(OpenGLUtils.entityData.getInstances() > 0)
+			OpenGLUtils.drawInstanced(OpenGLUtils.entityData, new Vec2f(Sprite.SIZE), scroll);
+		if(OpenGLUtils.itemData.getInstances() > 0)
+			OpenGLUtils.drawInstanced(OpenGLUtils.itemData, new Vec2f(Sprite.SIZE), scroll);
```
Comment (Bramelvix) [inline src/main/java/map/Level.java:322]: Put brackets around ifs pls

File: src/main/java/graphics/opengl/OpenGLUtils.java (hunk 1 above)
Comment (Bramelvix) [reply src/main/java/graphics/opengl/OpenGLUtils.java:289]: switches look nice enough

**Pull request merged**